utils/dom.js

```diff
@@ -150,14 +136,28 @@ export function getRectangleFromRange( range ) {
 		return range.getBoundingClientRect();
 	}
 
+	const { startContainer, startOffset } = range;
+
 	// If the collapsed range starts (and therefore ends) at an element node,
 	// `getClientRects` will return undefined. To fix this we can get the
-	// bounding rectangle of the element node to create a DOMRect based on that.
-	if ( range.startContainer.nodeType === ELEMENT_NODE ) {
-		const { x, y, height } = range.startContainer.getBoundingClientRect();
+	// selected child node and calculate a rectangle for that.
+	if ( startContainer.nodeType === ELEMENT_NODE ) {
+		const { childNodes } = startContainer;
+		const selectedNode = childNodes[
+			// Make sure there can be no errors.
```

Comment:
Were you able to determine why it's the case that sometimes `startOffset` exceeds the number of `childNodes` ?

Reply:
I've not encountered it. This is just to make sure it never does, if it would for some strange reason.

Reply:
Hmm, not entirely comfortable leaving this as an unknown.

Reply:
Hm, me neither. Let's remove this check and address any problems if they come up?

Reply:
Yeah, I think so. Will give it another round of testing this morning.

Reply:
The issue with removing the safety check can be seen when, with two paragraphs, you press ArrowLeft from the second paragraph (to the first).

Specifically, it can occur that `startOffset` is `1` even though the first paragraph only has a single child node (where `startContainer.childNodes[ 1 ]` is `undefined`). Based on the [documentation for `startOffset`](https://developer.mozilla.org/en-US/docs/Web/API/Range/startOffset), this might be expected ("number of child nodes between [...]"), where perhaps the caret is not considered as being within the text node, but rather after it?

While setting a bound with `Math.min` avoids the error, I don't think it's returning an accurate result. In the above scenario, it returns a bounding rect for the entire text node, effectively the same box as the entire paragraph, whereas with a collapsed selection at the end of the paragraph, we'd expect it to be zero width at an X coordinate aligned to the end of the paragraph.

I'm still finding 573269eccfdccfb85ddb8e2421bef7a82e6f1a22 to be the most durable approach, despite the fact that it requires manipulating the DOM to retrieve the rect. I wonder if we could exhaust all reasonable attempts to retrieve the rect before using the temporary `span` technique only if we still cannot get a valid rect (where invalid is either `undefined` or one with `0`'d property values).

Annoyingly, this appears to be a main point of browser inconsistency / bugginess. For example, in [this snippet](http://jsbin.com/lavixus/1/edit?html,js,output), adding a new line to the contenteditable causes the rect to become zero'd in latest Chrome, despite a handful of resolved Chromium bug reports which have claimed to have addressed these sorts of issues ([[1]](https://bugs.chromium.org/p/chromium/issues/detail?id=435438), [[2]](https://bugs.chromium.org/p/chromium/issues/detail?id=574363)).

Seeing modules like [rangefix](https://github.com/edg2s/rangefix) has me sad for the state of browser compatibility on these functions, and sadder that our end-to-end approach won't be able to test against a broad set of browsers. Nothing I've found seems to work perfectly, aside from [`bounding-client-rect`](https://github.com/webmodules/bounding-client-rect/blob/master/index.js) which uses basically the same approach as in 573269eccfdccfb85ddb8e2421bef7a82e6f1a22 . If it's a point of frequent regression across non-Chromium browsers that our E2E suite can't cover well, we could always extract these out these normalization functions into a separate project which uses a more thorough end-to-end approach (e.g. [Selenium](https://www.seleniumhq.org/)).

Reply:
TinyMCE obviously handles this well, where replacing this line:

https://github.com/WordPress/gutenberg/blob/049645b01766f679ac245d1674371b7f5aa3028c/blocks/rich-text/index.js#L662

...with:

```js
rect = this.editor.selection.getBoundingClientRect();
```

...will return the value we expect. Unsure if we can/want to be able to leverage this directly from our generic DOM utilities here.

Reply:
Yeah, this is what TinyMCE does:

https://github.com/tinymce/tinymce/blob/5f1e352a539359b351573d0621693e4580aa81d1/src/core/main/ts/caret/CaretPosition.ts#L126-L224

Maybe we can use this through `tinyMCE.dom.Selection`.

Reply:
We're also using TinyMCE's isEmpty check here.